### PR TITLE
Adding Integration Test Cases for Edge_Core (Issue #151)

### LINKED ISSUE #151: Integration test enhancement for DeviceTwin and edged
**What would you like to be added**:

More integration tests required for DeviceTwin and edged

**Why is this needed**:

DeviceTwin and Edged have various functionalities, and changes in one file may affect other functionality which depends on that file. Currently we do not have tests for many features/functionality of deviceTwin and edged. Only basic tests are present which do not test all the scenarios.

### PR DESCRIPTION
**Pull request by ShubhamM-15**
Added 6 test cases for Adding and updating devices.

**What type of PR is this?**
 /kind test

**What this PR does / why we need it**:
To improve and extend the Integration Test functionality
**Which issue(s) this PR fixes**:

Fixes #151

### CODE REVIEW

File: edge/test/integration/device/device_suite_test.go [hunk 1]
```diff
@@ -17,15 +17,14 @@ limitations under the License.
 package device_test
 
 import (
+	"github.com/kubeedge/kubeedge/edge/test/integration/utils/helpers"
```
Comment (sids-b) [inline edge/test/integration/device/device_suite_test.go]: Group golang, kubeedge and vendor packages together. As it was earlier

File: edge/test/integration/device/device_test.go [hunk 2]
```diff
@@ -18,19 +18,24 @@ package device_test
...
 	"encoding/json"
-
+	MQTT "github.com/eclipse/paho.mqtt.golang"
 	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
 	"github.com/kubeedge/kubeedge/edge/test/integration/utils/common"
 	. "github.com/kubeedge/kubeedge/edge/test/integration/utils/helpers"
-
-	MQTT "github.com/eclipse/paho.mqtt.golang"
 	_ "github.com/mattn/go-sqlite3"
```
Comment (sids-b) [inline edge/test/integration/device/device_test.go]: Group golang, kubeedge and vendor packages together. As it was earlier

File: edge/test/integration/device/device_test.go [hunk 3]
```diff
@@ -18,19 +18,24 @@ package device_test
...
 	. "github.com/onsi/gomega"
+	"net/http"
 )
 
 //getting deviceid from the DB and assigning to it
-var DeviceID string
+var DeviceID_N string
+var Device_N dttype.Device
+var DeviceID_WithAttr string
+var Device_ATT dttype.Device
```
Comment (sids-b) [inline edge/test/integration/device/device_test.go]: Camel case please,  instead of _

File: edge/test/integration/utils/helpers/helpers.go [hunk 4]
```diff
@@ -21,6 +21,7 @@ import (
 	"crypto/tls"
 	"database/sql"
 	"encoding/json"
+	"github.com/kubeedge/kubeedge/edge/test/integration/utils/edge"
```
Comment (sids-b) [inline edge/test/integration/utils/helpers/helpers.go]: Group golang, kubeedge and vendor packages together. As it was earlier

File: edge/test/integration/device/device_suite_test.go [hunk 5]
```diff
@@ -17,15 +17,16 @@ limitations under the License.
 package device_test
 
 import (
+
```
Comment (sids-b) [inline edge/test/integration/device/device_suite_test.go]: delete extra blank line

File: edge/test/integration/device/device_suite_test.go [hunk 6]
```diff
@@ -64,23 +65,36 @@ var (
 
 //Function to run the Ginkgo Test
 func TestEdgecoreEventBus(t *testing.T) {
+
```
Comment (sids-b) [inline edge/test/integration/device/device_suite_test.go]: delete extra line

File: edge/test/integration/utils/helpers/helpers.go [hunk 7]
```diff
@@ -51,6 +53,79 @@ type Device struct {
...
+	Type         string `json:"attr_type,omitempty"`
+	MetaData     string `json:"metadata,omitempty"`
+}
+
+func GenerateDeviceID(device_suffix string) string {
+	return device_suffix + edge.GetRandomString(10)
+}
+
+//Function to Generate Device
+func GenerateDevice(deviceID string, deviceName string, deviceState string) dttype.Device {
```
Comment (rohitsardesai83) [inline edge/test/integration/utils/helpers/helpers.go]: rename to CreateDevice

File: edge/test/integration/utils/helpers/helpers.go [hunk 8]
```diff
@@ -51,6 +53,79 @@ type Device struct {
...
+	var typemeta = dttype.TypeMetadata{Type: attributeType}
+	var attribute = dttype.MsgAttr{Value: attributeValue, Optional: &optional, Metadata: &typemeta}
+	device.Attributes[attributeName] = &attribute
+}
+
+//Function to add Twin attribute to existing device
+func AddTwinAttribute(device dttype.Device, attributeName string, attributeValue string, attributeType string) {
+	value := attributeValue
+	optional := true
+	value_meta := dttype.ValueMetadata{Timestamp: time.Now().Unix()}
```
Comment (rohitsardesai83) [inline edge/test/integration/utils/helpers/helpers.go]: follow camelCasing for variable names.

File: edge/test/integration/utils/helpers/helpers.go [hunk 9]
```diff
@@ -83,6 +158,79 @@ func GetDeviceStateFromDB(deviceID string) string {
...
+	for row.Next() {
+		err = row.Scan(&attribute.ID, &attribute.DeviceID, &attribute.Name, &attribute.Description, &attribute.Value, &attribute.Optional, &attribute.Type, &attribute.MetaData)
+		if err != nil {
+			common.Failf("Failed to scan DB rows: %v", err)
+		}
+		if string(attribute.DeviceID) == deviceID && attribute.Name == Name {
+			break
+		}
+	}
+
```
Comment (rohitsardesai83) [inline edge/test/integration/utils/helpers/helpers.go]: remove extra new lines throughout the code.

File: edge/test/integration/device/device_test.go [hunk 10]
```diff
@@ -18,8 +18,11 @@ package device_test
 
 import (
 	"encoding/json"
+	"net/http"
+
```
Comment (rohitsardesai83) [inline edge/test/integration/device/device_test.go]: remove extra new line.

File: edge/test/integration/utils/helpers/helpers.go (hunk 9 above)
Comment (ShubhamM-15) [reply edge/test/integration/utils/helpers/helpers.go]: I have added some new lines between two functions, to make it look better and segregated.

File: edge/test/integration/device/device_test.go (hunk 10 above)
Comment (ShubhamM-15) [reply edge/test/integration/device/device_test.go]: Imports are separated as go, kubeedge and vendor imports, new line is added to make it obvious. Siddharth earlier told me to make this segregation.

Comment (rohitsardesai83) [reply edge/test/integration/device/device_test.go]: one line is enough to separate the imports. ? Is the segregation not clear with one new line ?

Comment (ShubhamM-15) [reply edge/test/integration/device/device_test.go]: Oh. Got it. Thanks.
The image above displays only one line, so I got confused.
Removing the extra new line.

**Pull request merged**